src/Agent.Worker/Handlers/HandlerFactory.cs

```diff
@@ -0,0 +1,59 @@
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Handlers
+{
+    [ServiceLocator(Default = typeof(HandlerFactory))]
+    public interface IHandlerFactory : IAgentService
+    {
+        IHandler Create(
+            IExecutionContext executionContext,
+            HandlerData data,
+            Dictionary<string, string> inputs,
+            string taskDirectory);
+    }
+
+    public sealed class HandlerFactory : AgentService, IHandlerFactory
+    {
+        public IHandler Create(
+            IExecutionContext executionContext,
+            HandlerData data,
+            Dictionary<string, string> inputs,
+            string taskDirectory)
+        {
+            // Validate args.
+            Trace.Entering();
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(data, nameof(data));
+            ArgUtil.NotNull(inputs, nameof(inputs));
+            ArgUtil.NotNull(taskDirectory, nameof(taskDirectory));
+
+            // Create the handler.
+            IHandler handler;
+            if (data is NodeHandlerData)
+            {
+                // Node.
+                handler = HostContext.CreateService<INodeHandler>();
+                (handler as INodeHandler).Data = data as NodeHandlerData;
+            }
+            else if (data is ProcessHandlerData)
+            {
+                // Process.
+                handler = HostContext.CreateService<IProcessHandler>();
+                (handler as IProcessHandler).Data = data as ProcessHandlerData;
+            }
+            else
+            {
+                // This should never happen.
+                throw new NotSupportedException();
+            }
+
+            handler.ExecutionContext = executionContext;
```

Comment:
These properties are common to all handlers, so I exposed them on the handler base class and opted ahead and set them up here.